launchable/commands/record/build.py

```diff
@@ -169,100 +212,127 @@ def build(ctx: click.core.Context, build_name: str, source: List[str], max_days:
                     commit_hash = matched.group('hash')
                     name = matched.group('name')
                     if commit_hash and name:
-                        submodules.append((repo_name + "/" + name, repo_dist + "/" + name, commit_hash))
+                        r.append(Workspace(
+                            name=w.name + "/" + name,
+                            dir=w.dir + "/" + name,
+                            commit_hash=commit_hash))
+        return r
+
+    # figure out the commit hash and branch of those workspaces
+    def compute_hash_and_branch(ws: List[Workspace]):
+        # figure out w.commit_hash and w.branch
+        for w in ws:
+            try:
+                if not w.commit_hash:
+                    w.commit_hash = subprocess.check_output("git rev-parse HEAD".split(), cwd=w.dir).decode().replace("\n", "")
+            except Exception as e:
+                click.echo(
+                    click.style(
+                        "Can't get commit hash for {}. Do you run command under git-controlled directory? "
+                        "If not, please set a directory use by --source option.".format(w.dir),
+                        fg='yellow'),
+                    err=True)
+                print(e, file=sys.stderr)
+                sys.exit(1)
+            w.calc_branch_name()
+
+        # Our historical behaviour is to ignore --branch options. In that case, we should reject if that option is present
+        # I also think it's OK to allow this option to explicitly let the branch name set, since the inference
+        #  of the branch name from the workspace is inherently a little fragile
```

Comment:
Calling an attention to this one.

My understanding is that `--branch` option is originally developed as a companion add-on to `--commit`, where the idea is that we have to operate without any local git workspace.

Prior to refactoring, the `--branch` option appears to simply get ignored unless `--commit` option is also in use. But that is always a risky behaviour; Allowing user to set something, and simply ignoring it.

I'd like to suggest we allow the `--branch` option without `--commit` option. The idea is that, normally we infer the branch name automatically from the workspace, but that relies on some heuristics (i.e. picking up the first ref from `git show-ref`, so if multiple refs point to the same commit, only the luck would tell which one we get), so I can see the value of letting the user be precise.